src/shogun/base/library.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Sergey Lisitsyn, Sanuj Sharma
+ */
+
+#ifndef _LIBRARY_H_
+#define _LIBRARY_H_
+
+#include <shogun/base/manifest.h>
+#include <string_view>
+
+namespace shogun
+{
+
+    namespace internal
+    {
+        /** @brief
+         * Handles loading, calling and closing of plugins from shared object files.
+         */
+        class LibraryHandle;
+    }
+
+    /** @brief
+     * Provides an API for loading plugins as objects of this class
+     * and accessing Manifest of the loaded plugins.
+     * Uses LibraryHandle under the hood.
+     */
+    class Library
+    {
+    public:
+        /** Constructor to initialize library
+         * @param filename name of shared object file
+         */
+        Library(std::string_view filename);
+
+        /** Copy constructor
+         * @param other library object to be copied
+         */
+        Library(const Library& other);
+
+        /** Class Assignment operator
+         * @param other library object to be assigned
+         */
+        Library& operator=(const Library& other);
+
+        /** Equality operator
+         * @param first first Library
+         * @param second second Library
+         */
+        friend bool operator==(const Library& first, const Library& second);
+
+        /** Inequality operator
+         * @param first first Library
+         * @param second second Library
+         */
+        friend bool operator!=(const Library& first, const Library& second);
+
+        /** Destructor */
+        ~Library();
+
+        /** @return manifest of loaded library */
+        Manifest manifest();
+
+        /** @return name of function that accesses Manifest
+         * of loaded library.
+         */
+        static std::string_view get_manifest_accessor_name()
+        {
+            return kManifestAccessorName;
+        }
+
+        friend void unload_library(Library&& lib);
```

Comment:
why do you inject this to the outer namespace and have `void unload_library(Library&& lib);` 20 lines down?

Reply:
m_handle is private hence the friend part.